common/lockmanager_test.go

```diff
@@ -0,0 +1,362 @@
+
+    go func() {
+        defer key2LockUnlockBlock.Done();
+
+        // Lock key2 while key1 is already locked.
+        Lock(key2);
+
+        err := Unlock(key2);
+        if (err != nil) {
+            test.Fatalf("Failed to unlock key2 in the child thread.");
+        }
+    }();
+
+    // Wait for the thread to lock and unlock key2 before unlocking key1.
+    key2LockUnlockBlock.Wait();
+
+    err := Unlock(key1);
+    if (err != nil) {
+        test.Fatalf("Failed to unlock key1 in the main thread.");
+    }
+}
+
+// Lock(1) -> %Lock(1) -> Unlock(1) -> %Unlock(1).
+func TestMainThreadUnlockFirst(test *testing.T) {
+    clear();
+    defer clear();
+
+    key1 := "testkey1";
+    var threadLockBlock sync.WaitGroup;
+    var threadFinishBlock sync.WaitGroup;
+
+    threadLockBlock.Add(1);
+    threadFinishBlock.Add(1);
+    unlockInChildThreadFirst := make(chan struct{}, 1);
+
+    // Lock for the first time.
+    Lock(key1);
+
+    // This thread tries to lock key1 a second time but is blocked
+    // until key1 gets unlocked for the first time.
+    go func() {
+        defer threadFinishBlock.Done();
+
+        threadLockBlock.Done();
+
+        Lock(key1);
+
+        // Signal the child thread locked key1.
+        unlockInChildThreadFirst <- struct{}{};
+
+        err := Unlock(key1);
+        if (err != nil) {
+            test.Fatalf("Failed to unlock for the second time in the child thread.");
+        }
+    }();
+
+    // Wait for the second thread to try to lock key1
+    // while the lock is being used by the first thread.
+    threadLockBlock.Wait();
+    
+    // Small sleep to ensure the thread tries to lock key1.
+    time.Sleep(10 * time.Millisecond);
+
+    // Check if the child thread unlocked key1 before the main thread did.
+    if (len(unlockInChildThreadFirst) > 0) {
+        test.Fatalf("Failed to ensure the main thread unlocked key1 before the child thread unlocked it.");
+    }
+
+    // Unlock for the first time.
+    err := Unlock(key1);
+    if (err != nil) {
+        test.Fatalf("Failed to unlock for the first time in the main thread.");
+    }
+    
+    // Wait for the thread to lock and unlock key1.
+    threadFinishBlock.Wait();
+}
+
+// Lock(1) -> Unlock(1) -> %StaleCheck.
+func TestStaleRetentionWithNonStaleLock(test *testing.T) {
+    clear();
+    defer clear();
+
+    key1 := "testkey1";
+    var threadExecutionBlock sync.WaitGroup;
+
+    // Add key1 to the lockmap and have it unlocked.
+    Lock(key1);
+    Unlock(key1);
+
+    threadExecutionBlock.Add(1);
+
+    go func() {
+        defer threadExecutionBlock.Done();
+
+        RemoveStaleLocksOnce();
+    }();
+
+    // Wait for the thread to finish checking for staleness.
+    threadExecutionBlock.Wait();
+    
+    // Check if the stale lock got removed.
+    _, exists := lockMap.Load(key1);
+    if (!exists) {
+        test.Fatalf("Failed to retain lock even though it was accessed concurrently.");
+    }
+} 
+
+// Lock(1) -> %Stale Check.
```

Comment:
Maybe add something to this to indicate what you are doing with the timestamp.
Like: `Lock(1) -> MakeTimestampStale(1) -> %Stale Check.`.